### PR TITLE
[Merged by Bors] - Insert cleaned key phrases when map is empty

### PR DESCRIPTION
**Pull request by acrrd**
When we don't have any key phrase to return we insert again all cleaned one.

### CODE REVIEW

File: xayn-ai/src/coi/relevance.rs [hunk 1]
```diff
@@ -144,6 +145,7 @@ impl RelevanceMap {
         self.coi_to_relevance
             .remove(&coi_id)
             .map(|relevances| {
+                self.cleaned.remove(&coi_id);
```
Comment (janpetschexain) [inline xayn-ai/src/coi/relevance.rs:148]: is there a specific reason that this call is inside the closure? otherwise it could be called outside.

File: xayn-ai/src/coi/key_phrase.rs [hunk 2]
```diff
@@ -121,7 +123,8 @@ impl RelevanceMap {
                     (coi_id, relevance, penalized_relevance, key_phrase.clone())
                 })
             }).collect::<Vec<_>>();
-        relevances.sort_unstable_by(|(_, _, this, _), (_, _, other, _)| this.cmp(other).reverse());
+        relevances.sort_by(|(_, _, this, _), (_, _, other, _)| this.cmp(other).reverse());
```
Comment (acrrd) [inline xayn-ai/src/coi/key_phrase.rs:126]: To have consistency in tests. Does not really affect performance for us.

File: xayn-ai/src/coi/relevance.rs (hunk 1 above)
Comment (janpetschexain) [inline xayn-ai/src/coi/relevance.rs:148]: removing the id from the `cleaned` might starve this map soon. the main issue is that `RelevanceMap::remove()` is called in every iteration of the feedback loop (via `RelevanceMap::select_key_phrases()`), but while the key phrases of the corresponding id are removed & unified with new candidates & selected & inserted again in the `coi_to_relevance`/`relevance_to_key_phrase` maps, this is not the case yet for the `cleaned` map which will starve it. on the other hand, if we ran another iteration of the feedback loop, than we have new key phrases and don't have to worry about the `cleaned` anymore, is this the idea?

Comment (acrrd) [reply xayn-ai/src/coi/relevance.rs]: In the feedback loop we are removing the coi and put it back with updated data, so the coi is not clean anymore but we have fresh data in and the keywords that we just put in can be selected. 
After the remove the information about that coi in `cleaned` are stale.

File: xayn-ai/src/coi/relevance.rs [hunk 3]
```diff
@@ -169,6 +171,10 @@ impl RelevanceMap {
     pub(super) fn clean(&mut self, coi_id: CoiId, relevance: Relevance, key_phrase: &KeyPhrase) {
         if let Some(key_phrases) = self.relevance_to_key_phrase.get_mut(&(relevance, coi_id)) {
             key_phrases.retain(|this| this != key_phrase);
+            self.cleaned
+                .entry(coi_id)
+                .or_default()
+                .push((relevance, key_phrase.clone()));
```
Comment (janpetschexain) [inline xayn-ai/src/coi/relevance.rs:177]: it seems wasteful to clone the key phrase which we have just removed. instead of `key_phrases.retain()` we could find the index of the `key_phrase` in the `key_phrases` via `Iterator::position()` and `Vec::remove()` it and then `push()` it to the `cleaned` map.

this will also solve the remaining issue that even though the pushing is now inside the `if let`, it is still possible that we push a key phrase to `cleaned` which never existed in `key_phrases` before (the retaining will not remove anything if the `key_phrase` doesn't exist).

overall, we probably need more tests for the `RelevanceMap` methods to cover all those edge cases, maybe in a follow up pr.

File: xayn-ai/src/coi/relevance.rs [hunk 4]
```diff
@@ -251,6 +257,20 @@ impl RelevanceMap {
...
         }
     }
+
+    pub(super) fn insert_cleaned_if_empty(&mut self) {
+        if self.relevance_to_key_phrase.is_empty() {
+            let to_insert = std::mem::take(&mut self.cleaned);
+
+            for (coi_id, value) in to_insert.into_iter() {
+                // we need to insert in reverse order to keep the order in which
+                // the key phrases were originally
```
Comment (janpetschexain) [inline xayn-ai/src/coi/relevance.rs:267]: i double checked this, we could update the comment and point to the exact reason why we need to reverse. in `RelevanceMap::select_top_key_phrases()` we collect and sort the key phrases in descending order wrt the penalized relevances, then the selected key phrases are cleaned in that reversed order.

it's also important that this simple reverse only works if we keep up the invariant that the `penalty` must be sorted decreasingly. if we ever need to change that, then `cleaned` needs to also track the penalized relevances and reverse the insert order wrt them.

**Review by janpetschexain — APPROVED** [review]
👍

**Comment by acrrd** [comment]
bors merge
